### PR TITLE
Review

### PR DESCRIPTION
**Pull request by comzyh**

### CODE REVIEW

File: pinyin_marker.hpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+#include <iostream>
+#include <locale>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+#include <queue>
+#include <functional>
+namespace pinyin_marker {
```
Comment (Lcch) [inline pinyin_marker.hpp]: need blank line after include and namespace

File: aca.hpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
+#ifndef ACA_HPP
+#define ACA_HPP
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include <queue>
+
+template <typename CharT, typename DataT>
+struct TrieNode {
+  std::unordered_map<CharT, TrieNode<CharT, DataT> *> nexts;
```
Comment (Lcch) [inline aca.hpp]: std::unordered_map<CharT, TrieNode<CharT, DataT>*> nexts;

File: aca.hpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  TrieNode<CharT, DataT> * fail;
```
Comment (Lcch) [inline aca.hpp]: TrieNode<CharT, DataT> *fail;

File: aca.hpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  std::basic_string<CharT> str;
```
Comment (Lcch) [inline aca.hpp]: character(a CharT) instead?

File: aca.hpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  DataT data;
+  TrieNode () {
+    fail = nullptr;
+  }
+};
+
+template <typename CharT, typename DataT>
+struct Automation {
+  typedef TrieNode<CharT, DataT> NodeType;
+  NodeType * root_node = nullptr;
```
Comment (Lcch) [inline aca.hpp]: place the asterisk adjacent to either the type or to the variable name

File: aca.hpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      auto next = current->nexts.find(*chr);
+      if (next == current->nexts.end()) {
+        NodeType *new_node = new NodeType();
+        current->nexts.insert(std::make_pair(*chr, new_node));
+        current = new_node;
+      } else {
+        current = next->second;
+      }
+    }
+    current->str = str;
```
Comment (Lcch) [inline aca.hpp]: if only matching string has this field, it should remove it, because trienode has data field.

File: aca.hpp [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    callback(current->data); // callback, to allow user to modify data
+  }
+  void build() {
```
Comment (Lcch) [inline aca.hpp]: add a blank between methods

File: aca.hpp [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    std::queue<NodeType *> q;
```
Comment (Lcch) [inline aca.hpp]: asterisk problem: same above

File: aca.hpp [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    root_node->fail = root_node;
+    q.push(root_node);
+    while (!q.empty()) {
+      NodeType *h = q.front();
+      for (auto it=h->nexts.begin(); it != h->nexts.end(); it++ ) {
```
Comment (Lcch) [inline aca.hpp]: it = h->nexts.begin()

File: aca.hpp [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        NodeType * f = h->fail;
```
Comment (Lcch) [inline aca.hpp]: asterisk: same above

File: aca.hpp [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        const char32_t chr = it->first;
+        while(f->nexts.find(chr) == f->nexts.end() && f != root_node) {
+          f = f->fail;
+        }
+        if (f->nexts.find(chr) != f->nexts.end() && q.front() != root_node) {
+          h->nexts[chr]->fail = f->nexts.find(chr)->second;
+        } else {
+          h->nexts[chr]->fail = root_node;
+        }
+        q.push(h->nexts[chr]);
```
Comment (Lcch) [inline aca.hpp]: use it->second

File: aca.hpp [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+  }
+  void build() {
+    std::queue<NodeType *> q;
+    root_node->fail = root_node;
+    q.push(root_node);
+    while (!q.empty()) {
+      NodeType *h = q.front();
+      for (auto it=h->nexts.begin(); it != h->nexts.end(); it++ ) {
+        NodeType * f = h->fail;
+        const char32_t chr = it->first;
```
Comment (Lcch) [inline aca.hpp]: CharT instead of char32_t

File: aca.hpp [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        while(f->nexts.find(chr) == f->nexts.end() && f != root_node) {
```
Comment (Lcch) [inline aca.hpp]: add space between while and (

File: aca.hpp [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+          f = f->fail;
+        }
+        if (f->nexts.find(chr) != f->nexts.end() && q.front() != root_node) {
```
Comment (Lcch) [inline aca.hpp]: use h which is got already.

File: aca.hpp [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        }
+        q.push(h->nexts[chr]);
+      }
+      q.pop();
+    }
+  }
+  void find(const std::basic_string<CharT> &str, const std::function<void(size_t index, const DataT &data)> &callback) const {
+    NodeType * node = root_node;
+    for (size_t i = 0; i < str.length(); i++) {
+      char32_t chr = str[i];
```
Comment (Lcch) [inline aca.hpp]: CharT

File: aca.hpp [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+      while(node->nexts.find(chr) == node->nexts.end() && node != root_node) {
```
Comment (Lcch) [inline aca.hpp]: add space between while and (

**Review by Lcch — COMMENTED** [review]
aca.hpp to aca.h ?

**Comment by comzyh** [comment]
Outdated

**Pull request closed without merging**